Update release process (#1356)
cloudevents/spec: docs/GOVERNANCE.md

The specifications produced will adhere to the following:

> - The versioning scheme used will follow [semver](https://semver.org/).

Reviewer: Not blocking for this PR, but we could probably change all of this into the present tense now.

> - Specifications will be grouped into logical units and released at the same
>   time, with the same version number. This is true regardless of whether each

Reviewer: ```suggestion
- Specifications will be grouped into logical units, with all documents in a group released at the same
  time, with the same version number. This is true regardless of whether each
```

Author: done

  that the "minor" version number will always be zero and the `specversion`
  string will always be of the form `X.0`.

Note that these rules do not apply to unversioned documents, such as the
[documented extensions](../cloudevents/extensions/README.md).

To create a new release:

- Periodically the group will examine the list of extensions to determine
  if any action should be taken (e.g. removed due to it being stale). The
  creation of a new release will be the reminder to do this check. If any
  changes are needed then PRs will be created and reviewed by the group.

- For the most part we try to only use a single branch ("main") for our work.

>   Git branches will be used as a way to snapshot/tag releases. While we hope

Reviewer: I suggest we separate "the process" from "the motivation/reasoning".

I think it's worth calling out something like: "Each full (non-RC) release will have both a tag and a branch. The tag will be kept up-to-date with the tip of the branch. The purpose of having a branch is to support very minor fixes (typos, clarifications) which amend a release in place. The purpose of having a tag is to support GitHub releases, which can act as a notification channel for interested users."

Author: Copied your proposed text into the previous section - and deleted this paragraph

  to not push new commits to these branches, in practice, it is possible that
  changes to a release might be needed without "picking up" other changes that

>   might have been commited to "main". Creating the branch during the release

Reviewer: ```suggestion
  might have been committed to "main". Creating the branch during the release
```

Author: delete para

>   process will makes this easier to manage should the need should arise.

Reviewer: ```suggestion
  process makes this easier to manage should the need should arise.
```

Author: delete para

- Determine the new release version string. It should be of the form:
  `<subject>/vX.Y.Z`, e.g. `cloudevents/v1.0.4` or `subscriptions/v1.0.0`.

- Create a PR (for the "main" branch") that:
  - Modifies the repo's files to use the new version string appended with
    `-rc#` as appropriate. Make sure to remove all `-wip` suffixes as needed.
  - Merge the PR.
  - Initiate a final review/test of the release.

> - A "release candidate" tag will be created with the new release version

Reviewer: I thought we weren't going to use tags at all, on the grounds that we'd need to keep the tag up-to-date with the branch?

Author: yea, I'm not thrilled with the idea of dual maintenance once an RC is created. If the RC lives for a long time that's just going to be annoying. If you think having an RC tag is bad, or confusing, then I'm ok with just telling people to look at the tip of "main" for their review.

  string but with a suffix of `-rc#` (release candidate). This will indicate
  that the authors believe it is ready for final review/testing. This will be
  true for updates to existing specifications and for new specifications.

- When review/testing is completed, create a PR (for the "main" branch") that:

>   - Modifies the repo's files to use the new version string (w/o `-rc#`) as

Reviewer: ```suggestion
  - Modifies the repo's files to use the new version string (without `-rc#`) as
```

(I suspect this is easier to understand for readers whose primary language isn't English.)

Author: done__

    appropriate.
  - Update [RELEASES.md](RELEASES.md) to mention the new release, and
    reference the yet-to-be-created release tag.
  - Update the appropriate `*/RELEASE_NOTES.md` file with the changes
    for the release. The list can be generated via:
    `git log --pretty=format:%s main...v0.1 | grep -v "Merge pull"`

>     by replacing "v0.1" with the name of the previous release. Or, use github's

Reviewer: Probably best to give a fuller example, e.g. cloudevents/v1.0.4 here.

Author: done

    [new release](https://github.com/cloudevents/spec/releases/new) process

>     to generate the list w/o actually creating the release yet.

Reviewer: ```suggestion
    to generate the list without actually creating the release yet.
```

Author: done

- Merge the PR.
  - Note that the link checker should fail since any references to the new
    release tag will not be valid yet. This is expected.

> - Create a new branch with the same name as the new release version string

Reviewer: If we're keeping the tag up to date with the tip of the branch, do we actually need them to have different names? The purpose of that was primarily to disambiguate in URLs, right? (Maybe it's better to avoid the conflict anyway, but it's probably worth discussing.)

Reviewer: Resolving - that's fine.

  appended with `-branch` (e.g. `<subject>/vX.Y.Z-branch`).

>   - Use Github to create a

Reviewer: ```suggestion
  - Use GitHub to create a
```

    [new release](https://github.com/cloudevents/spec/releases/new).
    During that process, create a new tag with the new release version

>     string (e.g. `<subject>/vX.Y.Z`) w/o any suffix.

Reviewer: ```suggestion
    string (e.g. `<subject>/vX.Y.Z`) without any suffix.
```

>   - Rerun the github CI actions from the previous PR and the "main" branch as

Reviewer: ```suggestion
  - Rerun the GitHub CI actions from the previous PR and the "main" branch as
```

    they should all pass now; as a sanity check.

- Create an "announcement" highlighting the key features of the new release
  and any potential noteworthy activities of the group:
  - Send it to the mailing lists.
  - Announce the release on our [X account](http://x.com/cloudeventsio).
  - Add it to the "announcement" section of our
    [website](https://cloudevents.io/).

- If an update to a release is needed, create a PR for the appropriate
  branches (including "main"), and merge when ready. For any release that's

>   updated, you'll need to move the tag for that release to point to the HEAD

Reviewer: I *think* when referring to it normally, this should be head (or tip) rather than HEAD. I think HEAD is just the committish you can use in git commands.

```suggestion
  updated, you'll need to move the tag for that release to point to the head
```

Author: done

>   of that branch. We'll eventually setup a Github action to automatically do

Reviewer: ```suggestion
  of that branch. We'll eventually setup a GitHub action to automatically do
```

  it but for now you can do it via the CLI:

>   - `git pull` to make sure you have all latest branches and tags

Reviewer: ```suggestion
  - `git pull --tags` to make sure you have all latest branches and tags
```

>   - `git tag -d vX.Y.Z` to delete the old tag for the release

Reviewer: ```suggestion
  - `git tag -d <subject>/vX.Y.Z` to delete the old tag for the release
```

Author: done

>   - `git tag vX.Y.Z vX.Y.Z-branch` to create a new tag for the HEAD of the

Reviewer: ```suggestion
  - `git tag <subject>/vX.Y.Z <subject>/vX.Y.Z-branch` to create a new tag for the head of the
```

    release branch

>   - `git push REMOTE vX.Y.Z -f` to force the tag to updated in the github repo

Reviewer: ```suggestion
  - `git push REMOTE vX.Y.Z -f` to force the tag to updated in the GitHub repo
```

  up-to-date with the tip of the branch. The purpose of having a branch is to
  support very minor fixes (typos, clarifications) which amend a release in
  place. The purpose of having a tag is to support GitHub releases, which can
  act as a notification channel for interested users.

Note that these rules do not apply to unversioned documents, such as the
[documented extensions](../cloudevents/extensions/README.md).

To create a new release:

- Periodically the group will examine the list of extensions to determine
  if any action should be taken (e.g. removed due to it being stale). The
  creation of a new release will be the reminder to do this check. If any
  changes are needed then PRs will be created and reviewed by the group.

- Determine the new release version string. It should be of the form:
  `<subject>/vX.Y.Z`, e.g. `cloudevents/v1.0.4` or `subscriptions/v1.0.0`.

- Create a PR (for the "main" branch") that:
  - Modifies the repo's files to use the new version string appended with
    `-rc#` as appropriate. Make sure to remove all `-wip` suffixes as needed.
  - Merge the PR.
  - Initiate a final review/test of the release.

> - Create A "release candidate" with the new release version string but with a

Reviewer: ```suggestion
- Create a "release candidate" with the new release version string but with a
```

> - Create A "release candidate" with the new release version string but with a

Reviewer: It's not clear what "create a release candidate" actually means. What's the concrete step here? Should the "Create a PR" step above actually be a sub-bullet of this one?

Author: I'm reworking this section a bit and I think it'll address your concerns

  believe it is ready for final review/testing. This will be true for updates
  to existing specifications and for new specifications.

- When review/testing is completed, create a PR (for the "main" branch") that:
  - Modifies the repo's files to use the new version string (without `-rc#`)
    as appropriate..
  - Update [RELEASES.md](RELEASES.md) to mention the new release, and
    reference the yet-to-be-created release tag.
  - Update the appropriate `*/RELEASE_NOTES.md` file with the changes
    for the release. The list can be generated via:
    `git log --pretty=format:%s main...cloudevents/v1.0.3 | grep -v "Merge pull"`
    by replacing "cloudevents/v1.0.3" with the name of the previous release.
    Or, use GitHub's
    [new release](https://github.com/cloudevents/spec/releases/new) process
    to generate the list without actually creating the release yet.

- Merge the PR.
  - Note that the link checker should fail since any references to the new
    release tag will not be valid yet. This is expected.

- Create a new branch with the same name as the new release version string
  appended with `-branch` (e.g. `<subject>/vX.Y.Z-branch`).

>   - Use GitHub to create a

Author: RC Tag name:  <subject>@vX.Y.Z-rc1
Branch name:  <subject>@vX.Y.Z-branch
Tag name: <subject>@vX.Y.Z

Reviewer: But with the subject still, right?

Reviewer: Should this comment be included in the document somewhere? It might be useful just after "Each release will have both a tag and a branch".

Author: yup - just added

    [new release](https://github.com/cloudevents/spec/releases/new).
    During that process, create a new tag with the new release version

>     string (e.g. `<subject>/vX.Y.Z`) without any suffix.

Reviewer: ```suggestion
    string in the format `<subject>/vX.Y.Z` without any suffix.
```

(Or keep it parenthesized but remove the e.g., as it's not really an example.)

  place. The purpose of having a tag is to support GitHub releases, which can
  act as a notification channel for interested users.

- Naming will adhere to the following pattern:
  - Release Name: `SUBJECT@vX.Y.Z`
  - Release Candidate Tag Name: `SUBJECT@vX.Y.Z-rc#`
  - Release Branch Name: `SUBJECT@vX.Y.Z-branch`
  - Release Tag Name: `SUBJECT@vX.Y.Z`

Note that these rules do not apply to unversioned documents, such as the
[documented extensions](../cloudevents/extensions/README.md).

### Creating A New Release

- Periodically the group will examine the list of extensions to determine
  if any action should be taken (e.g. removed due to it being stale). The
  creation of a new release will be the reminder to do this check. If any
  changes are needed then PRs will be created and reviewed by the group.

- Determine the new release version string. It should be of the form:
  `SUBJECT@vX.Y.Z`, e.g. `cloudevents@v1.0.4` or `subscriptions@v1.0.0`.

- Before a new release is finalized, a "release candidate" (rc) should be
  created that identifies the versions of files in the repository that are to
  be reviewed. The process for a RC is as follows:

>   - Create a PR (for the "main" branch") that modifies the appropriate files

Reviewer: ```suggestion
  - Create a PR (for the "main" branch) that modifies the appropriate files
```

Author: Got it.
Thanks for the review and time!
